### ISSUE TITLE
Occur: Default action not working the first time

### ISSUE THREAD

**Issue by minad** [oantolin/embark#79]
There is a new issue, I found when testing #64 and #65.

1. M-x consult-line
2. M-SPC (this is embark-act-no-exit) or C-M-SPC (this is embark-act)
3. O (open occur)
4. RET (default action in occur buffer)

When pressing RET in the occur buffer, the Embark which-key opens and shows the default actions (embark-general-map). This happens only *the first time* I run the key sequence from above. When doing it again, everything behaves as expected.

**Comment by oantolin** [comment]
I figured this out! The entire embark library was not getting loaded until you ran embark-act the first time. This was too late to record the command that opened the minibuffer, `consult-line`. Then when you press `RET` in the embark occur buffer, `embark-default-action` doesn't know it is supposed to run `consult-line`! (It thinks the default action is `embark-act` actually, which is superconfusing). The second time, Embark is ready and records the command.

So, I'm not sure how to fix this.

I can documented that since embark-default-action needs to know the command that opened the minibuffer, it won't work unless Embark was already loaded before you ran said minibuffer-opening command. And I could recommend in the readme that you don't defer the loading.

**Comment by minad** [comment]
Ouch, this could have *occured* to me. Lazy loading :see_no_evil:

As for the fix: I think you should autoload the hook which records the command - it is not much code! This is not nice but would be correct.

I think in Emacs 28 there is also some new minibuffer-this-command variable which could probably be used instead of the record hook, so at that point one could stop adding the hook.

**Comment by minad** [comment]
I don't think I fully understand what you are doing here, but maybe it makes sense to have two different info caching functions, one for the minibuffer and one for regular buffers. Only the minibuffer version should be autoloaded. For regular buffers you don't need this, right, since the buffer still exists?

**Comment by oantolin** [comment]
> I think in Emacs 28 there is also some new minibuffer-this-command variable which could probably be used instead of the record hook, so at that point one could stop adding the hook.

Yes, maybe 8 years from now when it doesn't make sense to support Emacs 27 anymore. `marginalia--this-command` can be removed then too.

**Comment by minad** [comment]
>> I think in Emacs 28 there is also some new minibuffer-this-command variable which could probably be used instead of the record hook, so at that point one could stop adding the hook.

> Yes, maybe 8 years from now when it doesn't make sense to support Emacs 27 anymore. marginalia--this-command can be removed then too.

This was a fig leaf argument in order to justify the autoload hack. It is only needed for the soon-to-be-obsolete Emacs 27...

**Comment by oantolin** [comment]
I'll think about this a bit. I'm pretty sure I can by just recording `this-command` in the minibuffer, and waiting on the rest of the information. For `*Completions*` and Embark Occur buffers I do need the full `embark--cache-info`, but not for the minibuffer.

**Comment by oantolin** [comment]
If I could get by with recording just `this--command` in the `minibuffer-setup-hook` (and just the last `(minibuffer-contents)` in the `minibuffer-exit-hook`, for the resume functionality), would you suggest I autoload them?

**Comment by minad** [comment]
I would probably autoload a small code block with the small cache functions and the hook setup. But now I start to wonder, is there really no other better way to obtain the command which opened the minibuffer? Some call stack inspection?

**Comment by oantolin** [comment]
I'll have to restructure a bit, because now several places assume more information is cached in the minibuffer. But it can be pared down to just the `embark--command`.

**Comment by oantolin** [comment]
You know, a cheap partial work around would be to use `marginalia--this-command` if bound, since we recommend that package not be deferred. 😛

**Comment by oantolin** [comment]
Maybe it's more polite to put all of these information caching hooks in a global minor mode.

**Comment by minad** [comment]
@oantolin I also thought about this, but you must load the mode in order to activate it. And then you can just directly require embark if you put the mode in embark.el.

**Comment by oantolin** [comment]
Ah, right. At any rate it seems easy to change the minibuffer caching to make it very minimal.

**Comment by minad** [comment]
You could also split embark.el and put the mode in a separate file. I think this would be the purest most idiomatic way of doing this. Everything which must be loaded immediately should go into this file, the lazy stuff in another file. But I am not proposing this seriously, I would only do this if you start introducing splits. Not only for this. (I know you won't like it ;)

**Comment by oantolin** [comment]
OK, done. I think the caching strategy is much more sensible now. I was trying to treat minibuffers, embark occur buffers and completions buffers the same, but really have slightly different needs.
